Add `--xdebug` option in php-wasm cli

mho22 opened the pull request:
## Motivation for the change, related issues

This pull request adds a `--xdebug` option in :

- `php-wasm/cli`
- `wp-playground/cli`

[Roadmap](https://github.com/WordPress/wordpress-playground/issues/2315)

## Implementation details

For PHP-WASM : 

- Simple implementation of a Xdebug option when checking for a `--xdebug` argument.

For WP-Playground : 

- Addition of `ASYNCIFY_ONLY` functions called.
- Simple implementation of a Xdebug option when checking for a `xdebug` option.
 

## Testing Instructions


`php/xdebug.php`

```php
<?php

$test = 42; // Set a breakpoint on 42 and run Debugger;

echo "Output!\n";

function test() {
	echo "Hello Xdebug World!\n";
}

test();
```

For PHP-WASM : 

`@php-wasm/cli` is required.

```
PHP=7.2 node_modules/.bin/cli --xdebug --version
```

```
PHP 7.2.34-dev (cli) (built: Jul  8 2025 17:39:20) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Xdebug v3.1.7-dev, Copyright (c) 2002-2022, by Derick Rethans
```

```
node_modules/.bin/cli --xdebug php/xdebug.php


> Breaks on $test = 42
```

For WP-Playground : 

`@wp-playground/cli` is required.

```
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI({ php: '7.2', command: "server", xdebug: true });

const result = await cliServer.playground.run({ code: "<?php echo 'With Xdebug v' . phpversion( 'xdebug' );" });

console.log(result.text);
```

```
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
Booted!
Running the Blueprint...
Running the Blueprint – 100%
Finished running the blueprint
WordPress is running on http://127.0.0.1:53090
With Xdebug v3.1.7-dev
```

```
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI({ command: "server", xdebug: true });

const result = await cliServer.playground.run({ scriptPath: `${process.cwd()}/php/xdebug.php` });

console.log(result.text);
```

```
node scripts/cli.js
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
Booted!
Running the Blueprint...
Running the Blueprint – 100%
Finished running the blueprint
WordPress is running on http://127.0.0.1:53133


> Breaks on $test = 42
```

mho22 commented:
It couldn't be that easy. wp-playground cli crashes for PHP versions below `8.2` with error : 

```
Error [RuntimeError]: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (wasm://wasm/php.wasm-07829542:wasm-function[7771]:0x6a3e4f)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-07829542:wasm-function[7373]:0x6771f7)
```

As these functions are already in `ASYNCIFY_ONLY` I compiled PHP `8.1` with `WITH_DEBUG=yes` to see if this could be more verbose but it makes the script run successfully instead.

mho22 commented:
I faced a Heisenbug here. Changing `-g2` from : 

```
# Preserve symbol names in node.js build – the bundle size doesn't matter as much
# as on the web, and this makes debugging **much** easier.
if [ "$EMSCRIPTEN_ENVIRONMENT" = "node" ]; then \
	echo -n ' -g2 ' >> /root/.emcc-php-wasm-flags; \
fi; \
```

with `-g3` and no more error occur. So : 


`-O3 -g3` and `-O0 -g2` options association do make the script pass. Not easy to debug.

Anyway, that's the following stack trace when running `node --stack-trace-limit=100 cripts/cli.js` 

```                                                       
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
Error [RuntimeError]: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
    at php.wasm.execute_ex
    ... 56 lines matching cause stack trace ...
    at callUserCallback 
    at Timeout._onTimeout
    at listOnTimeout
    at process.processTimers  {
  cause: Error
      at Asyncify.handleSleep 
      at _emscripten_sleep 
      at php.wasm.__wrap_select
      at php.wasm.Curl_socket_check
      at php.wasm.curl_multi_perform
      at php.wasm.zif_curl_exec
      at php.wasm.execute_internal
      at xdebug.so.xdebug_execute_internal
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex 
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_INCLUDE_OR_EVAL_SPEC_OBSERVER_HANDLER
      at php.wasm.ZEND_USER_OPCODE_SPEC_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_INCLUDE_OR_EVAL_SPEC_OBSERVER_HANDLER
      at php.wasm.ZEND_USER_OPCODE_SPEC_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex
      at php.wasm.ZEND_INCLUDE_OR_EVAL_SPEC_OBSERVER_HANDLER 
      at php.wasm.ZEND_USER_OPCODE_SPEC_HANDLER
      at php.wasm.execute_ex
      at xdebug.so.xdebug_execute_ex 
      at php.wasm.zend_execute 
      at php.wasm.zend_execute_scripts 
      at php.wasm.dynCall_iiiii 
      at ret.<computed> [as dynCall_iiiii] 
      at invoke_iiiii 
      at php.wasm.php_execute_script
      at php.wasm.dynCall_ii 
      at ret.<computed> [as dynCall_ii] 
      at invoke_ii 
      at php.wasm.wasm_sapi_handle_request 
      at ret.<computed> 
      at t.wasmExports.<computed> 
      at Object.doRewind 
      at Timeout.poll
      at listOnTimeout
      at process.processTimers
}
```

I added the following new functions in `ASYNCIFY_ONLY_UNPREFIXED` : 

```diff
+ _emscripten_sleep
+ execute_internal
+ ZEND_INCLUDE_OR_EVAL_SPEC_CONST_HANDLER
+ ZEND_INCLUDE_OR_EVAL_SPEC_TMPVAR_HANDLER
```

But it isn't enough

adamziel commented:
Let's debug php 8.2 separately from adding the cli switch. Let's just say it's experimental in the option description

adamziel commented:
Ah it's not 8.2 - it's below 8.2

mho22 commented:
Yep, unfortunately it concerns every php version below 8.2

adamziel commented:
`__wrap_select` maybe?

adamziel commented:
Also, would `-s STACK_SIZE=50MB` change anything? Also, try rebasing on top of the latest trunk, I've made some changes in socket and error handling internals

mho22 commented:
Unfortunately, setting `STACK_SIZE` and rebasing didn't help, and `__wrap_select` is already included in `ASYNCIFY_ONLY_UNPREFIXED`. I'm continuing to investigate.

adamziel commented:
Would changing ASYNCIFY_ONLY to ASYNCIFY_ADD change anything?

mho22 commented:
I replaced `STACK_SIZE` with `ASYNCIFY_STACK_SIZE`, replaced `ASYNCIFY_ONLY` with `ASYNCIFY_ADD` and I still get the same stack trace. I added `-sASSERTIONS=1` and it returned two new `warnings` but I don't know if this is related: 

```
node scripts/cli.js
```
```
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
warning: unsupported syscall: __syscall_socketpair

warning: unsupported syscall: __syscall_setsockopt

Error [RuntimeError]: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (wasm://wasm/php.wasm-094adf5e:wasm-function[7771]:0x98a8e7)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-094adf5e:wasm-function[7373]:0x93854b)
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (wasm://wasm/php.wasm-094adf5e:wasm-function[7771]:0x98a44a)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-094adf5e:wasm-function[7373]:0x93854b)
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (wasm://wasm/php.wasm-094adf5e:wasm-function[7771]:0x98a44a)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-094adf5e:wasm-function[7373]:0x93854b)
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER (wasm://wasm/php.wasm-094adf5e:wasm-function[7771]:0x98a44a) {
  cause: Error
      at Asyncify.handleSleep (file:///Users/mho/Work/Projects/Development/Web/Professional/xdebug/template/node_modules/@php-wasm/node/index.js:339345:45)
      at returnCallback (file:///Users/mho/Work/Projects/Development/Web/Professional/xdebug/template/node_modules/@php-wasm/node/index.js:335905:51)
      at wasm_poll_socket (file:///Users/mho/Work/Projects/Development/Web/Professional/xdebug/template/node_modules/@php-wasm/node/index.js:335912:12)
      at imports.<computed> (file:///Users/mho/Work/Projects/Development/Web/Professional/xdebug/template/node_modules/@php-wasm/node/index.js:326098:22)
      at php.wasm.__wrap_select (wasm://wasm/php.wasm-094adf5e:wasm-function[9140]:0xac810f)
      at php.wasm.Curl_socket_check (wasm://wasm/php.wasm-094adf5e:wasm-function[21814]:0x120da4e)
      at php.wasm.curl_multi_perform (wasm://wasm/php.wasm-094adf5e:wasm-function[21779]:0x12055c6)
      at php.wasm.zif_curl_exec (wasm://wasm/php.wasm-094adf5e:wasm-function[925]:0x4058ad)
      at php.wasm.execute_internal (wasm://wasm/php.wasm-094adf5e:wasm-function[7352]:0x934634)
      at wasm://wasm/00127ec6:wasm-function[244]:0xd25a
```

`at wasm://wasm/00127ec6:wasm-function[243]:0xceee` represents `xdebug.so.xdebug_execute_ex` but

mho22 commented:
I replaced `STACK_SIZE` with `ASYNCIFY_STACK_SIZE`, replaced `ASYNCIFY_ONLY` with `ASYNCIFY_ADD` and I still get the same stack trace. I added `-sASSERTIONS=1` and it returned two new `warnings` but I don't know if this is related: 

```
node scripts/cli.js
```
```
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
warning: unsupported syscall: __syscall_socketpair

warning: unsupported syscall: __syscall_setsockopt

Error [RuntimeError]: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
    at php.wasm.execute_ex
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
    at php.wasm.execute_ex
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
    at php.wasm.execute_ex
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
{
  cause: Error
      at Asyncify.handleSleep 
      at returnCallback 
      at wasm_poll_socket 
      at imports.<computed> 
      at php.wasm.__wrap_select 
      at php.wasm.Curl_socket_check 
      at php.wasm.curl_multi_perform
      at php.wasm.zif_curl_exec 
      at php.wasm.execute_internal
      at wasm://wasm/00127ec6:wasm-function[244]:0xd25a
```

adamziel commented:
Would it break with just a sleep(1) call? Or fopen/fread? Or is it only for curl?

mho22 commented:
It simply breaks because of `xdebug: true` in : 

```javascript
import { runCLI } from "@wp-playground/cli";

await runCLI({ php: '8.1', xdebug: true });
```

I do not call any sleep or curl, only these two lines. It works as expected when using the `@php-wasm/cli` but something is  breaking here.

adamziel commented:
Aha! These two lines do a lot, like installing WordPress. Let's isolate the issue

mho22 commented:
Thanks for the insights, I dug deeper and found out the error occurred when running `installWordpress()` function from `boot.ts` file in `packages/playground/cli/src/worker-thread.ts` at line `142`. 

I added a new option in `runCLIargs` named `dataSqlPath` to avoid having to call the `installWordpress()` method [ Because `installWordpress()` is called when that option is undefined ] : 

`boot.ts` line `180` : 
```javascript
if (!options.dataSqlPath) {
	if (!(await isWordPressInstalled(php))) {
		await installWordPress(php);
	}

	if (!(await isWordPressInstalled(php))) {
		throw new Error('WordPress installation has failed.');
	}
}
```

 and my script stopped successfully at breakpoint for every PHP versions. All of this because it never ran `installWordpress()`. 


So this is my new script : 

```javascript
import { withPHPIniValues } from '@php-wasm/universal';
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI({ php: '7.2', dataSqlPath: 'none', xdebug: true });

const result = await cliServer.playground.run( { scriptPath : `${process.cwd()}/php/xdebug.php` } );

console.log( result.text );

await cliServer.playground.request({
	url: '/wp-admin/install.php?step=2',
	method: 'POST',
	body: {
		language: 'en',
		prefix: 'wp_',
		weblog_title: 'My WordPress Website',
		user_name: 'admin',
		admin_password: 'password',
		// The installation wizard demands typing the same password twice
		admin_password2: 'password',
		Submit: 'Install WordPress',
		pw_weak: '1',
		admin_email: 'admin@localhost.com',
	},
})
```

This code will break on `$test = 42` for every PHP versions and for PHP versions below 8.2 it will crash when running `withPHPIniValues`.

Another thing I noticed : Two requests are appearing in `CALL STACK` for PHP versions above 8.1 while it crashes for the others.

<img width="1715" height="1061" alt="Capture d’écran 2025-07-15 à 10 13 06" src="https://github.com/user-attachments/assets/b342ade9-8412-4f41-9ee7-f1a103433283" />


New stack when running `node scripts/cli.js` with PHP `8.1`

```
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.wordpress.org/release/wordpress-6.8.2-RC1.zip
Fetching SQLite integration plugin...
Booting WordPress...
Booted!
Running the Blueprint...
Running the Blueprint – 100%
Finished running the blueprint
WordPress is running on http://127.0.0.1:64123
Output!
Hello Xdebug World!

Error [RuntimeError]: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
    at php.wasm.execute_ex
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER
    at php.wasm.execute_ex
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
    at php.wasm.execute_ex 
    at wasm://wasm/00127ec6:wasm-function[243]:0xceee
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER{
  cause: Error
      at Asyncify.handleSleep 
      at _emscripten_sleep 
      at php.wasm.__wrap_select
      at php.wasm.Curl_socket_check 
      at php.wasm.curl_multi_perform
      at php.wasm.zif_curl_exec
      at php.wasm.execute_internal 
      at wasm://wasm/00127ec6:wasm-function[244]:0xd25a
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
      at php.wasm.execute_ex
```

adamziel commented:
What if we took WordPress out of the equation entirely and called php.runStream() directly? Or went through php wasm cli, not playground cli? What would be the simplest script that crashes it? Would it involve curl? Or would regular sleep or fread be enough?

mho22 commented:
I think I'm onto something interesting! I cloned the wordpress directory into the root of my project and am currently step-debugging through the entire WordPress installation script.

<img width="1724" height="1072" alt="Capture d’écran 2025-07-15 à 11 32 58" src="https://github.com/user-attachments/assets/6afe6c14-5d30-461a-ae41-8b28ab4c11bf" />

This should probably allow us to detect a lot more `unreachable` functions from within the project. Maybe.

@adamziel I haven’t yet managed to reproduce a minimal crashing example.

adamziel commented:
I reproduced the crash with this script:

```php
<?php

// Create a multi handle
$multiHandle = curl_multi_init();

// Initialize individual curl handles
$handles = [];

// First request to WordPress.org API
$ch1 = curl_init();
curl_setopt($ch1, CURLOPT_URL, 'https://api.wordpress.org/core/version-check/1.7/');
curl_setopt($ch1, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch1, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch1, CURLOPT_SSL_VERIFYPEER, false);
$handles['version_check'] = $ch1;
curl_multi_add_handle($multiHandle, $ch1);

// Second request to WordPress.org main site
$ch2 = curl_init();
curl_setopt($ch2, CURLOPT_URL, 'https://wordpress.org/');
curl_setopt($ch2, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch2, CURLOPT_FOLLOWLOCATION, true);
curl_setopt($ch2, CURLOPT_SSL_VERIFYPEER, false);
$handles['main_site'] = $ch2;
curl_multi_add_handle($multiHandle, $ch2);

// Execute all requests
$running = null;
do {
    curl_multi_exec($multiHandle, $running);
    curl_multi_select($multiHandle);
} while ($running > 0);

// Collect results
$results = [];
foreach ($handles as $key => $handle) {
    $results[$key] = curl_multi_getcontent($handle);
    curl_multi_remove_handle($multiHandle, $handle);
    curl_close($handle);
}

// Close multi handle
curl_multi_close($multiHandle);

// Display results
echo "WordPress.org API Response:\n";
echo $results['version_check'] . "\n\n";

echo "WordPress.org Main Site (first 200 chars):\n";
echo substr($results['main_site'], 0, 200) . "...\n";

```

And this CLI command

```
PHP=7.4 node  --stack-trace-limit=100 --loader=./packages/meta/src/node-es-module-loader/loader.mts ./packages/php-wasm/cli/src/main.ts ./test.php --xdebug
```

adamziel commented:
```php
<?php
sleep(1);
```

This is enough to crash it – it seems like anything async does, which makes me think it's about asyncifying xdebug functions, not native PHP functions.

mho22 commented:
Aha! Thank you Adam! But I think you found another issue, since `sleep(1)` is also crashing with PHP8.4 while it is not crashing for my issue. Anyways, let's try to solve it first then.

mho22 commented:
The following stack trace show a `table index is out of bounds` when I add `WITH_DEBUG` during Xdebug compilation : 

```
PHP=8.1 node --stack-trace-limit=1000 node_modules/.bin/cli php/sleep.php --xdebug

RuntimeError: table index is out of bounds
    at xdebug.so.xdebug_execute_internal
    at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
    ... 20 lines matching cause stack trace ...
    at callUserCallback 
    at Timeout._onTimeout 
    at listOnTimeout (node:internal/timers:608:17)
    at process.processTimers (node:internal/timers:543:7) {
  cause: Error
      at Asyncify.handleSleep 
      at _emscripten_sleep
      at php.wasm.wasm_sleep
      at php.wasm.zif_sleep
      at php.wasm.execute_internal
      at xdebug.so.xdebug_execute_internal
      at php.wasm.ZEND_DO_FCALL_SPEC_OBSERVER_HANDLER 
      at php.wasm.execute_ex 
      at xdebug.so.xdebug_execute_ex 
      at php.wasm.zend_execute 
      at php.wasm.zend_execute_scripts 
      at php.wasm.dynCall_iiiii
      at ret.<computed> [as dynCall_iiiii] 
      at invoke_iiiii 
      at php.wasm.php_execute_script 
      at php.wasm.dynCall_ii 
      at ret.<computed> [as dynCall_ii] 
      at invoke_ii 
      at php.wasm.do_cli 
      at php.wasm.dynCall_iii 
      at ret.<computed> [as dynCall_iii] 
      at invoke_iii 
      at php.wasm.main
      at php.wasm.run_cli 
      at ret.<computed> 
      at t.wasmExports.<computed>
      at Object.doRewind
      at Timeout.poll [as _onTimeout] 
      at listOnTimeout
      at process.processTimers 
}

Node.js v23.11.0
```

adamziel commented:
I've built xdebug with `-O0 -g2` and that uncovered a few more functions in the stacktrace:

* `xdebug_execute_ex`
* `xdebug_execute_internal`

```
RuntimeError: unreachable
    at php.wasm.ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER (wasm://wasm/php.wasm-075e022a:wasm-function[7653]:0x681989)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-075e022a:wasm-function[7256]:0x656855)
    ... 19 lines matching cause stack trace ...
    at file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:18482:39
    at callUserCallback (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:8775:4)
    at Timeout._onTimeout (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:16840:4)
    at listOnTimeout (node:internal/timers:608:17)
    at process.processTimers (node:internal/timers:543:7) {
  cause: Error
      at Asyncify.handleSleep (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:35599:41)
      at _emscripten_sleep (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:16848:19)
      at php.wasm.wasm_sleep (wasm://wasm/php.wasm-075e022a:wasm-function[8851]:0x6f2ed9)
      at php.wasm.zif_sleep (wasm://wasm/php.wasm-075e022a:wasm-function[4375]:0x510a01)
      at php.wasm.execute_internal (wasm://wasm/php.wasm-075e022a:wasm-function[7237]:0x65478e)
      at xdebug.so.xdebug_execute_internal (wasm://wasm/xdebug.so-0048d842:wasm-function[248]:0x1b0da)
      at php.wasm.ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER (wasm://wasm/php.wasm-075e022a:wasm-function[7653]:0x681658)
      at php.wasm.execute_ex (wasm://wasm/php.wasm-075e022a:wasm-function[7256]:0x656855)
      at xdebug.so.xdebug_execute_ex (wasm://wasm/xdebug.so-0048d842:wasm-function[247]:0x1ac59)
      at php.wasm.zend_execute (wasm://wasm/php.wasm-075e022a:wasm-function[7258]:0x656d71)
      at php.wasm.zend_execute_scripts (wasm://wasm/php.wasm-075e022a:wasm-function[6790]:0x631f09)
      at php.wasm.dynCall_iiiii (wasm://wasm/php.wasm-075e022a:wasm-function[39376]:0xfc621c)
      at ret.<computed> [as dynCall_iiiii] (file:///Users/cloudnik/www/Automattic/core/plugins/pla
```


I've put them in XDebug's `ASYNCIFY_ADD` param. I also noticed that `execute_internal` is not listed in the PHP ASYNCIFY_ONLY list so I've added it and rebuilt both XDebug and PHP.

That got me further and I got a `Table index out of bounds` error which reminded me of @zaerl's PR (#2290). I've set `-s STACK_SIZE=5MB` and got a new stacktrace with more `xdebug` functions:

```
      at xdebug.so.xdebug_fd_read_line_delim (wasm://wasm/xdebug.so-0048d842:wasm-function[716]:0xa5dbd)
      at xdebug.so.xdebug_dbgp_cmdloop (wasm://wasm/xdebug.so-0048d842:wasm-function[698]:0x9c2c4)
      at xdebug.so.xdebug_dbgp_init (wasm://wasm/xdebug.so-0048d842:wasm-function[694]:0x9bb08)
      at xdebug.so.xdebug_init_debugger (wasm://wasm/xdebug.so-0048d842:wasm-function[632]:0x879de)
      at xdebug.so.xdebug_debug_init_if_requested_at_startup (wasm://wasm/xdebug.so-0048d842:wasm-function[639]:0x889fa)
      at xdebug.so.xdebug_execute_user_code_begin (wasm://wasm/xdebug.so-0048d842:wasm-function[251]:0x1bef6)
      at xdebug.so.xdebug_execute_ex 
```

I'm rebuilding XDebug again with those functions listed.

mho22 commented:
Alright we are on the same page! To be honest, I thought it wasn't necessary to list Asyncify functions in xdebug. But it seems we do : 

That's the first list I made when experimenting Xdebug Asyncify a couple weeks ago : 

```diff
+ zm_post_zend_deactivate_xdebug
+ xdebug_debugger_post_deactivate
+ xdebug_dbgp_deinit
+ xdebug_execute_begin
+ xdebug_execute_user_code_begin
+ xdebug_init_debugger
+ xdebug_debug_init_if_requested_at_startup
+ xdebug_dbgp_init
+ xdebug_execute_ex
+ xdebug_statement_call
+ xdebug_debugger_statement_call
+ xdebug_dbgp_breakpoint
+ xdebug_dbgp_cmdloop
+ xdebug_fd_read_line_delim
```

Maybe this will help you save time.

adamziel commented:
@mho22 Thank you! I've rebuilt with the functions that came up in the stack trace and with that list and still no bueno. Well, at least we're stuck further down the road!

adamziel commented:
aha, I added `execute_internal` to asyncify_import_list in xdebug Dockerfile and it worked. I'll push my changes in a moment.

mho22 commented:
Briliant! But what gave you the idea to add `execute_internal` into `-sASYNCIFY_IMPORTS`?

adamziel commented:
I glanced at this stack trace and thought "hm, it seems like `xdebug_execute_internal` calls `execute_internal` – would that make it an import? let's try it". I'm not sure why we don't need to mark `execute_ex` as an import, though.

```
      at Asyncify.handleSleep (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:32238:49)
      at _emscripten_sleep (file:///Users/cloudnik/www/Automattic/core/plugins/playground/packages/php-wasm/node/asyncify/php_8_0.js:15612:23)
      at php.wasm.__wrap_select (wasm://wasm/php.wasm-09555cba:wasm-function[9788]:0x86566a)
      at php.wasm.Curl_poll (wasm://wasm/php.wasm-09555cba:wasm-function[24188]:0xe126bf)
      at php.wasm.Curl_multi_wait (wasm://wasm/php.wasm-09555cba:wasm-function[24118]:0xe0ad17)
      at php.wasm.curl_multi_wait (wasm://wasm/php.wasm-09555cba:wasm-function[24117]:0xe0a464)
      at php.wasm.zif_curl_multi_select (wasm://wasm/php.wasm-09555cba:wasm-function[1006]:0x3ef5b8)
      at php.wasm.execute_internal (wasm://wasm/php.wasm-09555cba:wasm-function[8078]:0x78e38a)
      at xdebug.so.xdebug_execute_internal (wasm://wasm/xdebug.so-0048d91e:wasm-function[248]:0x1b0fe)
      at php.wasm.ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER (wasm://wasm/php.wasm-09555cba:wasm-function[8499]:0x7bfafe)
      at php.wasm.execute_ex (wasm://wasm/php.wasm-09555cba:wasm-function[8097]:0x7909fb)
      at xdebug.so.xdebug_execute_ex (wasm://wasm/xdebug.so-0048d91e:wasm-function[247]:0x1ac59)
      at php.wasm.ZEND_DO_FCALL_SPEC_RETVAL_UNUSED_HANDLER (wasm://wasm/php.wasm-09555cba:wasm-function[8499]:0x7bf967)
      at php.wasm.execute_ex (wasm://wasm/php.wasm-09555cba:wasm-function[8097]:0x7909fb)
      at xdebug.so.xdebug_execute_ex (wasm://wasm/xdebug.so-0048d91e:wasm-function[247]:0x1ac59)
      at php.wasm.zend_execute (wasm://wasm/php.wasm-09555cba:wasm-function[8099]:0x791340)
      at php.wasm.zend_execute_scripts (wasm://wasm/php.wasm-09555cba:wasm-function[7617]:0x75f65b)
```

adamziel commented:
I just pushed a version where this test script works on PHP 8.0:

```
<?php
// 
function testFopen() {
    echo "fopen\n";
    $response = fopen('https://api.wordpress.org/core/version-check/1.7/', 'r');
    echo strlen(fread($response, 1024)) . "\n";
    echo "\n";
    fclose($response);
}

function testFileGetContents() {
    echo "file_get_contents\n";
    echo strlen(file_get_contents('https://api.wordpress.org/core/version-check/1.7/')) . "\n";
    echo "\n";
}

function testSingleCurl() {
    echo "Single cURL handle request\n";
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, 'https://api.wordpress.org/core/version-check/1.7/');
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, false);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    echo "\n";

    if ($response === false) {
        echo "cURL Error: " . curl_error($ch) . "\n";
    } else {
        echo "Single cURL Response (HTTP $httpCode):\n";
        echo strlen($response) . "\n";
    }

    curl_close($ch);
}

function testMultiCurl() {
    echo "Multi cURL handle request\n";

    // Create a multi handle	
    $multiHandle = curl_multi_init();

    // Initialize individual curl handles
    $handles = [];

    // First request to WordPress.org API
    $ch1 = curl_init();
    curl_setopt($ch1, CURLOPT_URL, 'https://api.wordpress.org/core/version-check/1.7/');
    curl_setopt($ch1, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch1, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch1, CURLOPT_SSL_VERIFYPEER, false);
    $handles['version_check'] = $ch1;
    curl_multi_add_handle($multiHandle, $ch1);

    // Second request to WordPress.org main site
    $ch2 = curl_init();
    curl_setopt($ch2, CURLOPT_URL, 'https://wordpress.org/');
    curl_setopt($ch2, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch2, CURLOPT_FOLLOWLOCATION, true);
    curl_setopt($ch2, CURLOPT_SSL_VERIFYPEER, false);
    $handles['main_site'] = $ch2;
    curl_multi_add_handle($multiHandle, $ch2);

    // Execute all requests
    $running = null;
    do {
        curl_multi_exec($multiHandle, $running);
        curl_multi_select($multiHandle);
    } while ($running > 0);

    // Collect results
    $results = [];
    foreach ($handles as $key => $handle) {
        $results[$key] = curl_multi_getcontent($handle);
        curl_multi_remove_handle($multiHandle, $handle);
        curl_close($handle);
    }

    // Close multi handle
    curl_multi_close($multiHandle);

    // Display results
    echo "WordPress.org API Response:\n";
    echo strlen($results['version_check']) . "\n\n";

    echo "WordPress.org Main Site (first 200 chars):\n";
    echo strlen(substr($results['main_site'], 0, 200)) . "...\n";
}

// Call all the functions
testFopen();
testFileGetContents();
testSingleCurl();
testMultiCurl();
```

I'll back off from this PR for now. I hope that helps! LMK if you run into any problems with other PHP versions.

Btw, we should run some of the asyncify unit tests (not all, that would be too much) with and without xdebug loaded, e.g. `networking`, `proc_open`, maybe something else – just to make sure we're checking popular failure modes on all PHP versions.

mho22 commented:
Thanks a lot for your help! 

> Btw, we should run some of the asyncify unit tests (not all, that would be too much) with and without xdebug loaded, e.g. networking, proc_open, maybe something else – just to make sure we're checking popular failure modes on all PHP versions.

Of course. I'll test every file with xdebug locally and consider those that are the most crucial as new CI tests.

mho22 commented:
@adamziel I tried a hundred times to run your test file with : 

```
PHP=8.0 node  --stack-trace-limit=100 --loader=./packages/meta/src/node-es-module-loader/loader.mts ./packages/php-wasm/cli/src/main.ts ./test.php --xdebug
```

without success...

I then tried to run that command without the `--xdebug` option and it also returned a `unreachable` error.

I don't know how you made it and at this point I am too afraid to ask.

Edit : PHP8.1 is valid for @wp-playground/cli and @php-wasm/cli using your test file.

adamziel commented:
I wonder if I somehow pushed the wrong files – I couldn't run it using the binaries from this PR either. I've just rebuilt 8.0 locally and it worked well. I've just pushed those recent changes.

mho22 commented:
That's intriguing... I also rebuilt it and it didn't work. I need to be more careful I guess.

adamziel commented:
Weird! I wonder what went wrong 🤔 does it work with the current version of this branch?

mho22 commented:
Yes, but somehow my previous version that worked for `@wp-playground/cli` does not work anymore. Anyway I continue my investigation.

adamziel commented:
Oh, interesting! Baby steps I guess. I've rebuilt 7.4 in the meantime just to see if the test script would work and it does 🤞

mho22 commented:
I found out your script works for PHP versions below PHP 8.2 but they crash with `@wp-playground/cli`. I should make a markdown table to note what I tested : 

| PHP | php-wasm/cli    | wp-playground/cli |
| ---- | -------- | ------- |
| 8.4  | ✅ | ✅ |
| 8.3  | ✅ | ✅ |
| 8.2  | ✅ | ✅ |
| 8.1  | ✅ |  ✅ |
| 8.0  | ✅ | ✅ |
| 7.4  | ✅ | ✅ |
| 7.3  | ✅ | ✅ |
| 7.2  | ✅ | ✅ |

adamziel commented:
I've added two more items to asyncify_list and got `@wp-playground/cli` to boot on 7.4 – hopefully it's down to rebuilding and finding function names 🤞

mho22 commented:
@adamziel Have you tried to run a script like the following for PHP7.4 ?

```javascript
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI( { php : '7.4', xdebug : true } );

const result = await cliServer.playground.run( { scriptPath : `${process.cwd()}/test.php` } );

console.log( result.text );
```

`test.php` being your above script.

adamziel commented:
<img width="2376" height="2396" alt="CleanShot 2025-07-16 at 01 09 03@2x" src="https://github.com/user-attachments/assets/32c83b20-b43d-415f-9246-dfedcfaba73b" />

mho22 commented:
Ok, something seems to be wrong with my setup. Nothing is working on my end, while everything builds smoothly on yours. Sorry for taking up your time. But it surely helps 😄.

mho22 commented:
This is my output when I run : 

`node --stack-trace-limit=100 scripts/cli.js` 

```
Starting a PHP server...
Setting up WordPress undefined
Resolved WordPress release URL: https://downloads.w.org/release/wordpress-6.8.2.zip
Fetching SQLite integration plugin...
Booting WordPress...
Error [RuntimeError]: unreachable
    at php.wasm.zval_try_get_string_func (wasm://wasm/php.wasm-071342ce:wasm-function[6519]:0x604673)
    at php.wasm.zval_try_get_string_func (wasm://wasm/php.wasm-071342ce:wasm-function[6519]:0x604569)
    at php.wasm.zend_include_or_eval (wasm://wasm/php.wasm-071342ce:wasm-function[7980]:0x691311)
    at php.wasm.ZEND_INCLUDE_OR_EVAL_SPEC_TMPVAR_HANDLER (wasm://wasm/php.wasm-071342ce:wasm-function[7981]:0x6918af)
    at php.wasm.ZEND_USER_OPCODE_SPEC_HANDLER (wasm://wasm/php.wasm-071342ce:wasm-function[8339]:0x6b5911)
    at php.wasm.execute_ex (wasm://wasm/php.wasm-071342ce:wasm-function[7518]:0x6607da)
    at xdebug.so.xdebug_execute_ex (wasm://wasm/xdebug.so-008432ae:wasm-function[243]:0x1d99e)
    at php.wasm.ZEND_INCLUDE_OR_EVAL_SPEC_CV_HANDLER (wasm://wasm/php.wasm-071342ce:wasm-function[7982]:0x691fc2)
```

After rebuilding PHP7.4

and using my script : 

```javascript 
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI( { php : '7.4', xdebug : true } );

const result = await cliServer.playground.run( { scriptPath : `${process.cwd()}/php/test.php` } );

console.log( result.text );
```

based on `npm run local-package-repository`

mho22 commented:
I reset my setup and your version works for me locally now. This is sorcery.

adamziel commented:
I wonder if nx cache got in the way during rebuilding (`nx reset`) 🤔

mho22 commented:
That's exactly what I suspected. I'll try something, but I'll focus on PHP versions above 8.1 after.

Edit: I couldn't reproduce the issue, let's move on.

mho22 commented:
After a few calibration with the different Dockerfiles, I found out only `execute_internal` was necessary in `xdebug` Dockerfile to run your script with every PHP versions. It means we no more need `ASYNCIFY_ADD`. Great isn't it.

mho22 commented:
I am now looking to find a way to conduct tests specifically for xdebug without creating new test files.

mho22 commented:
The following test files are now tested with `xdebug` enabled/disabled : 

```
php-file-get-contents.spec.ts
php-fopen.spec.ts
php-fsockopen.spec.ts
php-networking.spec.ts
php.spec.ts
```

About `php.spec.ts`, it uses a `beforeEach` function where it loads the PHP runtime. That's where I can enable Xdebug or not. So, to avoid having to run every tests with or without Xdebug I added a skip option to every `describe` we don't target. In summary, every tests other than `popen` and `proc_open` have a `{ skip : option.withXdebug }` option that will skip if xdebug is enabled.

Unfortunately, two `proc_open` tests  also needed that skip :

```
feof() returns true when exhausted the synchronous data
feof() returns true when exhausted the synchronous data
```

Because it sometimes returned : 

```diff
- resource(3) of type (stream),
+ resource(1) of type ((null)),
```

adamziel commented:
Good work here @mho22! I am away this week but I'll review in details on Monday.

> Unfortunately, two proc_open tests also needed that skip :
> 
> feof() returns true when exhausted the synchronous data
> feof() returns true when exhausted the synchronous data
> Because it sometimes returned :
> 
> - resource(3) of type (stream),
> + resource(1) of type ((null)),

So the process resource is unusable with xdebug loaded? That sounds like a bug. By "sometimes" do you mean it's non-deterministic? Or can it be reliably reproduced? Either way, let's log it separately and not block this PR on it - xdebug is optional and asyncify is still and added value.

adamziel commented:
Actually it's a reasonably small pr that also looks good. Thank you!

mho22 commented:
Yes it wasn't a huge PR after all for sure. Let's update https://github.com/WordPress/wordpress-playground/issues/314

mho22 commented:
> So the process resource is unusable with xdebug loaded? That sounds like a bug. By "sometimes" do you mean it's non-deterministic? Or can it be reliably reproduced? Either way, let's log it separately and not block this PR on it - xdebug is optional and asyncify is still and added value.

This is non-deterministic. I faced errors for PHP versions above 8.0 sometimes and then errors for every versions. I think we'll need proper attention on these tests in another issue.
